File: bindings/ocaml/lib/safe_network.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe network operations for IP address validation and classification. *)

type ipv4 = {
  a: int;
  b: int;
  c: int;
  d: int;
}

let ipv4_to_string ip =
  Printf.sprintf "%d.%d.%d.%d" ip.a ip.b ip.c ip.d

let parse_ipv4 address =
  match String.split_on_char '.' address with
  | [a; b; c; d] ->
      (try
        let a = int_of_string a in
        let b = int_of_string b in
        let c = int_of_string c in
        let d = int_of_string d in
        if a >= 0 && a <= 255 &&
           b >= 0 && b <= 255 &&
           c >= 0 && c <= 255 &&
           d >= 0 && d <= 255
        then Some { a; b; c; d }
        else None
      with Failure _ -> None)
  | _ -> None

let is_valid_ipv4 address =
  match parse_ipv4 address with
  | Some _ -> true
  | None -> false

let is_private address =
  match parse_ipv4 address with
  | None -> false
  | Some ip ->
      (* 10.0.0.0/8 *)
      ip.a = 10 ||
      (* 172.16.0.0/12 *)
      (ip.a = 172 && ip.b >= 16 && ip.b <= 31) ||
      (* 192.168.0.0/16 *)
      (ip.a = 192 && ip.b = 168)

let is_loopback address =
  match parse_ipv4 address with
  | None -> false
  | Some ip -> ip.a = 127

let is_public address =
  is_valid_ipv4 address && not (is_private address) && not (is_loopback address)

let format_ipv4 = ipv4_to_string

File: bindings/julia/src/SafeNetwork.jl
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeNetwork

Safe network operations for IP address validation and classification.
"""
module SafeNetwork

export IPv4, parse_ipv4, is_valid_ipv4, is_private, is_loopback, is_public, format_ipv4

"""
    IPv4

Represents an IPv4 address.
"""
struct IPv4
    a::UInt8
    b::UInt8
    c::UInt8
    d::UInt8
end

"""
    parse_ipv4(address::AbstractString) -> Union{IPv4, Nothing}

Parse an IPv4 address string.
"""
function parse_ipv4(address::AbstractString)::Union{IPv4, Nothing}
    parts = split(address, ".")
    length(parts) != 4 && return nothing

    octets = UInt8[]
    for part in parts
        n = tryparse(Int, part)
        n === nothing && return nothing
        (n < 0 || n > 255) && return nothing
        push!(octets, UInt8(n))
    end

    IPv4(octets[1], octets[2], octets[3], octets[4])
end

"""
    is_valid_ipv4(address::AbstractString) -> Bool

Check if a string is a valid IPv4 address.
"""
function is_valid_ipv4(address::AbstractString)::Bool
    parse_ipv4(address) !== nothing
end

"""
    is_private(address::AbstractString) -> Bool

Check if an IPv4 address is in a private range.
"""
function is_private(address::AbstractString)::Bool
    ip = parse_ipv4(address)
    ip === nothing && return false

    # 10.0.0.0/8
    ip.a == 10 && return true
    # 172.16.0.0/12
    ip.a == 172 && ip.b >= 16 && ip.b <= 31 && return true
    # 192.168.0.0/16
    ip.a == 192 && ip.b == 168 && return true

    false
end

"""
    is_loopback(address::AbstractString) -> Bool

Check if an IPv4 address is a loopback address (127.0.0.0/8).
"""
function is_loopback(address::AbstractString)::Bool
    ip = parse_ipv4(address)
    ip === nothing && return false
    ip.a == 127
end

"""
    is_public(address::AbstractString) -> Bool

Check if an IPv4 address is public (not private or loopback).
"""
function is_public(address::AbstractString)::Bool
    is_valid_ipv4(address) && !is_private(address) && !is_loopback(address)
end

"""
    format_ipv4(ip::IPv4) -> String

Format an IPv4 address as a string.
"""
function format_ipv4(ip::IPv4)::String
    "$(ip.a).$(ip.b).$(ip.c).$(ip.d)"
end

end # module
